#AISV_801 Natural Language Processing #

## Instructor Joseph Meyer ##

Lab4: Explore Word2vec

Bobby Wen

Dataset: poem_sentiment

Exploration

    Choose a word, find the n most similar words
    
    Code the famous king - man = queen

Download Poem_sentiment

    Create a function to calculate the average word2vec of a given string

### Load required libraries ###

In [ ]:
!pip install datasets

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from datasets import load_dataset

### load and check dataset ###

In [ ]:
# #Load poem_sentiment dataset
dataset = load_dataset("poem_sentiment")

In [ ]:
### Check Dataset
print (dataset)

DatasetDict({
    train: Dataset({
        features: ['id', 'verse_text', 'label'],
        num_rows: 892
    })
    validation: Dataset({
        features: ['id', 'verse_text', 'label'],
        num_rows: 105
    })
    test: Dataset({
        features: ['id', 'verse_text', 'label'],
        num_rows: 104
    })
})


In [ ]:
### train test split on just data is not needed, by could be used if needed
# Using the train_test_split function, we are able to create training and testing data set from a single data set
#X_train, X_test, y_train, y_test = train_test_split(dataset["train"]['verse_text'], dataset["train"]['label'])

In [ ]:
### Split the training and testing datasets as intended by the dataset
X_train, y_train = dataset["train"]['verse_text'], dataset["train"]['label']
X_test, y_test = dataset["test"]['verse_text'], dataset["test"]['label']

In [ ]:
import random
print (random.choice(X_train))
print (len(X_train), "Records")

the head that lay against your knees
892 Records


In [ ]:
print (random.choice(y_train))
print (len(y_train), "Records")

1
892 Records


In [ ]:
print (random.choice(X_test))
print (len(X_test), "Records")

indignantly i hurled the cry:
104 Records


In [ ]:
print (random.choice(y_test))
print (len(y_test), "Records")

2
104 Records


In [ ]:
###  Class reference version

# import gensim

# def load_word2vec_model(model_path):
#     # Load the Word2Vec model
#     model = gensim.models.KeyedVectors.load_word2vec_format(model_path, binary=True)
#     return model

# def find_similar_words(word, model, topn=5):
#     # Find words similar to the given word
#     similar_words = model.most_similar(word, topn=topn)
#     return similar_words

# def word_vector(word, model): # Get the vector representation of a word
#     return model[word]


In [ ]:
import logging
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

### Gensim - “Generate Similar” ###

Gensim is a Python library for topic modelling, document indexing and similarity retrieval with large corpora. Target audience is the natural language processing (NLP) and information retrieval (IR) community. It provides access to Word2Vec and other word embedding algorithms for training, and it also allows pre-trained word embeddings that can be downloaded from the internet to be loaded.

Why Gensim?
Super fast
The fastest library for training of vector embeddings – Python or otherwise. The core algorithms in Gensim use battle-hardened, highly optimized & parallelized C routines.

** Data Streaming **
Gensim can process arbitrarily large corpora, using data-streamed algorithms. There are no "dataset must fit in RAM" limitations.

** Platform independent **
Gensim runs on Linux, Windows and OS X, as well as any other platform that supports Python and NumPy.

** Proven **
With thousands of companies using Gensim every day, over 2600 academic citations and 1M downloads per week, Gensim is one of the most mature ML libraries.

** Open source **
All Gensim source code is hosted on Github under the GNU LGPL license, maintained by its open source community. For commercial arrangements, see Business Support.

** Ready-to-use models and corpora **
The Gensim community also publishes pretrained models for specific domains like legal or health, via the Gensim-data project.

In [ ]:
import gensim.downloader
# Show all available models in gensim-data
print(list(gensim.downloader.info()['models']))

['fasttext-wiki-news-subwords-300', 'conceptnet-numberbatch-17-06-300', 'word2vec-ruscorpora-300', 'word2vec-google-news-300', 'glove-wiki-gigaword-50', 'glove-wiki-gigaword-100', 'glove-wiki-gigaword-200', 'glove-wiki-gigaword-300', 'glove-twitter-25', 'glove-twitter-50', 'glove-twitter-100', 'glove-twitter-200', '__testing_word2vec-matrix-synopsis']


** Models **


name	num vectors 	file size	base dataset	description


conceptnet-numberbatch-17-06-300	1917247	1168 MB	ConceptNet, word2vec, GloVe, and OpenSubtitles 2016	ConceptNet Numberbatch consists of state-of-the-art semantic vectors (also known as word embeddings) that can be used directly as a representation of word meanings or as a starting point for further machine learning. ConceptNet Numberbatch is part of the ConceptNet open data project. ConceptNet provides lots of ways to compute with word meanings, one of which is word embeddings. ConceptNet Numberbatch is a snapshot of just the word embeddings. It is built using an ensemble that combines data from ConceptNet, word2vec, GloVe, and OpenSubtitles 2016, using a variation on retrofitting.


fasttext-wiki-news-subwords-300	999999	958 MB	Wikipedia 2017, UMBC webbase corpus and statmt.org news dataset (16B tokens)	1 million word vectors trained on Wikipedia 2017, UMBC webbase corpus and statmt.org news dataset (16B tokens).


glove-twitter-100	1193514	387 MB	Twitter (2B tweets, 27B tokens, 1.2M vocab, uncased)	Pre-trained vectors based on 2B tweets, 27B tokens, 1.2M vocab, uncased (https://nlp.stanford.edu/projects/glove/)

glove-twitter-200	1193514	758 MB	Twitter (2B tweets, 27B tokens, 1.2M vocab, uncased)	Pre-trained vectors based on 2B tweets, 27B tokens, 1.2M vocab, uncased (https://nlp.stanford.edu/projects/glove/).

glove-twitter-25	1193514	104 MB	Twitter (2B tweets, 27B tokens, 1.2M vocab, uncased)	Pre-trained vectors based on 2B tweets, 27B tokens, 1.2M vocab, uncased (https://nlp.stanford.edu/projects/glove/).

glove-twitter-50	1193514	199 MB	Twitter (2B tweets, 27B tokens, 1.2M vocab, uncased)	Pre-trained vectors based on 2B tweets, 27B tokens, 1.2M vocab, uncased (https://nlp.stanford.edu/projects/glove/)

glove-wiki-gigaword-100	400000	128 MB	Wikipedia 2014 + Gigaword 5 (6B tokens, uncased)	Pre-trained vectors based on Wikipedia 2014 + Gigaword 5.6B tokens, 400K vocab, uncased (https://nlp.stanford.edu/projects/glove/).

glove-wiki-gigaword-200	400000	252 MB	Wikipedia 2014 + Gigaword 5 (6B tokens, uncased)	Pre-trained vectors based on Wikipedia 2014 + Gigaword, 5.6B tokens, 400K vocab, uncased (https://nlp.stanford.edu/projects/glove/).

glove-wiki-gigaword-300	400000	376 MB	Wikipedia 2014 + Gigaword 5 (6B tokens, uncased)	Pre-trained vectors based on Wikipedia 2014 + Gigaword, 5.6B tokens, 400K vocab, uncased (https://nlp.stanford.edu/projects/glove/).

glove-wiki-gigaword-50	400000	65 MB	Wikipedia 2014 + Gigaword 5 (6B tokens, uncased)	Pre-trained vectors based on Wikipedia 2014 + Gigaword, 5.6B tokens, 400K vocab, uncased (https://nlp.stanford.edu/projects/glove/).

word2vec-google-news-300	3000000	1662 MB	Google News (about 100 billion words)	Pre-trained vectors trained on a part of the Google News dataset (about 100 billion words). The model contains 300-dimensional vectors for 3 million words and phrases. The phrases were obtained using a simple data-driven approach described in 'Distributed Representations of Words and Phrases and their Compositionality' (https://code.google.com/archive/p/word2vec/).



word2vec-ruscorpora-300	184973	198 MB	Russian National Corpus (about 250M words)	Word2vec Continuous Skipgram vectors trained on full Russian National Corpus (about 250M words). The model contains 185K words.




GloVe stands for Global Vectors for Word Representation.

GloVe is an unsupervised learning algorithm for obtaining vector representations for words. Training is performed on aggregated global word-word co-occurrence statistics from a corpus, and the resulting representations showcase interesting linear substructures of the word vector space.

I am loading a pre-trained models to not have to train our dataset.

A small data set glove-twitter-25 was used initially to test code.  Later larger data sets were tested.  Finally the largest dataset is used to test the famous King - man prediction.

We use the pre-train model to find words similar to 'king'

In [ ]:
# Download the "glove-twitter-25" embeddings
#w2v_model = gensim.downloader.load('glove-twitter-25')   ###  smallest data set to test code
#w2v_model = gensim.downloader.load('glove-wiki-gigaword-300')   ### Larger pre-trained model to test accuracy
w2v_model = gensim.downloader.load('fasttext-wiki-news-subwords-300')   ### Larger pre-trained model to maximize accuracy

[==================================================] 100.0% 958.5/958.4MB downloaded


In [ ]:
# Use the downloaded vectors as usual:
w2v_model.most_similar('king', topn=5)

[('king-', 0.7838029861450195),
 ('boy-king', 0.7704817652702332),
 ('queen', 0.7704246640205383),
 ('prince', 0.7700967192649841),
 ('kings', 0.7668929696083069)]

Code the famous king - man = queen test

In [ ]:
vec = w2v_model['king'] - w2v_model['man']

In [ ]:
w2v_model.similar_by_vector(vec, topn=5)

[('king', 0.4447827935218811),
 ('kings', 0.36625999212265015),
 ('Cináed', 0.36057743430137634),
 ('co-king', 0.35103851556777954),
 ('Wuyue', 0.3504401445388794)]

In [ ]:
vec = w2v_model['king'] - w2v_model['man'] + w2v_model['woman']

In [ ]:
w2v_model.similar_by_vector(vec, topn=5)

[('king', 0.8185327053070068),
 ('queen', 0.7570335865020752),
 ('queen-mother', 0.6896809339523315),
 ('king-', 0.6787333488464355),
 ('queen-consort', 0.6555202603340149)]

In [ ]:
w2v_model.most_similar(vec, topn=5)

[('king', 0.8185327053070068),
 ('queen', 0.7570335865020752),
 ('queen-mother', 0.6896809339523315),
 ('king-', 0.6787333488464355),
 ('queen-consort', 0.6555202603340149)]

In [ ]:
w2v_model.most_similar(positive=['king', 'woman'], negative=['man'], topn=5)

[('queen', 0.7786749005317688),
 ('queen-mother', 0.7143871784210205),
 ('king-', 0.6981282234191895),
 ('queen-consort', 0.6724597811698914),
 ('monarch', 0.6666999459266663)]

### Gensim 2 model, saving and reloading test to check if model will work on next tier ###

In [ ]:
from gensim.test.utils import common_texts
from gensim.models import Word2Vec
from gensim.models import KeyedVectors

# # Store just the words + their trained embeddings.
w2v_model.save("word2vec1.wordvectors")

In [ ]:
# Load back with memory-mapping = read-only, shared across processes.
wv = KeyedVectors.load("word2vec1.wordvectors", mmap='r')

In [ ]:
### test of the reloaded model
vector = wv['computer']  # Get numpy vector of a word
print (vector)

[-0.089791   -0.017821   -0.037268   -0.013295   -0.0090625  -0.03825
 -0.026104   -0.089276    0.051467   -0.0021857  -0.0074708  -0.016283
 -0.056232    0.026517   -0.040413    0.011108    0.1075      0.05117
  0.038878   -0.0071759  -0.027288    0.052936    0.0048894   0.053154
 -0.012498   -0.0036237  -0.031653    0.0080834  -0.018473    0.048122
  0.0098432  -0.014012    0.0041448  -0.039648    0.03404     0.012708
  0.02619    -0.010029    0.06783    -0.028244   -0.068825   -0.11998
 -0.034457    0.014159    0.019905   -0.045616    0.057433   -0.039549
 -0.002202   -0.0027748  -0.012895    0.01229     0.011554    0.017499
 -0.054615    0.058455    0.033276    0.019049   -0.03536    -0.034284
  0.014129   -0.014776    0.11628     0.057433    0.045229   -0.0088706
  0.018039    0.023135   -0.034525    0.024178    0.012592    0.013363
  0.032109    0.016538   -0.0022877  -0.027777    0.025676   -0.019492
  0.025357   -0.038128   -0.028559   -0.011518   -0.043368    0.037497
 -0.0228

In [ ]:
wv.most_similar(vector, topn=2)

[('computer', 1.0000001192092896), ('computers', 0.8472651243209839)]

In [ ]:
#model = w2v_model
vec1 = wv['airplane']
vec2 = wv['wings']
vec3 = wv['boat']
vec4 = wv['water']

In [ ]:
result_vec = vec1 - vec2 + vec4
print(result_vec)

[ 4.48516011e-02  1.77494735e-02  7.00179907e-03  2.13823602e-01
  1.42352998e-01  9.28856954e-02 -2.20099986e-02 -8.93699974e-02
  1.00948006e-01  2.01800186e-03 -4.16420028e-02 -1.61290038e-02
  6.18999917e-03 -2.25130022e-02  4.42000031e-02  3.72980013e-02
  9.74790007e-02  4.33389992e-02 -2.26499960e-02  2.24609990e-02
  1.98272001e-02 -5.42418584e-02  4.16069999e-02  1.57020055e-02
  5.17360009e-02 -7.53870010e-02 -6.99819950e-03 -1.45260012e-02
  6.85199909e-03  1.14156596e-01 -3.56000103e-03  1.62874997e-01
  1.69504017e-01 -1.26860011e-02  6.13770001e-02 -3.24536003e-02
  6.20739982e-02 -5.01598008e-02 -1.03100576e-03 -5.56529015e-02
  3.53326052e-02 -1.79419994e-01 -3.04399990e-02  9.51298745e-04
  8.08179975e-02 -6.09169975e-02 -2.53149942e-02 -1.24399969e-03
 -5.65499999e-02  1.04581006e-02  8.44020024e-02 -1.09930001e-02
  7.96940029e-02 -8.91719982e-02 -3.00120041e-02  3.85790020e-02
 -1.54500082e-03 -1.60820037e-02  3.93669978e-02  7.51660019e-03
  8.26187953e-02  4.22400

In [ ]:
wv.similar_by_vector(result_vec, topn=5)

[('water', 0.6646305322647095),
 ('water--', 0.5573056936264038),
 ('drinking-water', 0.5420796871185303),
 ('waste-water', 0.5359698534011841),
 ('potable', 0.5343053936958313)]

In [ ]:
wv.most_similar(result_vec, topn=5)  ### same output as similar_by_vector

[('water', 0.6646305322647095),
 ('water--', 0.5573056936264038),
 ('drinking-water', 0.5420796871185303),
 ('waste-water', 0.5359698534011841),
 ('potable', 0.5343053936958313)]

### 2. Average word2vec of a given string  ###

    Download Poem_sentiment and create a function to calculate the average word2vec of a given string

In [ ]:
### This function can take a sentence and split it for processing
def take_sentence_and_split_into_array(sentence):
  """
  Reads a sentence from the user and splits it into an array.

  Returns:
    The sentence, stored in an array.
  """
  print(sentence)
  words = sentence.split(" ")
  array = []

  for word in words:
    array.append(word)

  return array

In [ ]:
### This function allow the user to input their own sentence for processing
def read_sentence_and_split_into_array():
  """
  Reads a sentence from the user and splits it into an array.

  Returns:
    The sentence, stored in an array.
  """
  sentence = input("Enter a sentence: ")
  words = sentence.split(" ")
  array = []

  for word in words:
    array.append(word)

  return array

In [ ]:
# This function calculates the average word2vec of a text array
import numpy as np
#from sklearn.metrics.pairwise import cosine_similarity

def average_word2vec(text_list):
    model = w2v_model   ### hardcode of model to be used from setup
    test = []

##loop the given sentence
    for word in text_list:
        try:
            word_embeding = model.get_vector(word, norm=True)
            test.append(np.average(word_embeding,axis=0))
        except KeyError:
            continue

    return test

In [ ]:
### take a random sentence from the test data set loaded from poem_sentiment
sentence = random.choice(X_test)

sentence_array = take_sentence_and_split_into_array(sentence)
print(sentence_array)
print('number of words in sentence', len(sentence_array))

she sought for flowers
['she', 'sought', 'for', 'flowers']
number of words in sentence 4


In [ ]:
### Print the average word2vec of input sentence
avg_word2vec = average_word2vec(sentence_array)
print ('Average word2vec of the string', avg_word2vec)

Average word2vec of the string [-0.0034214917, -0.0009899102, -0.002198077, 7.521182e-05]


### Input your own phrase ###

In [ ]:
sentence_array = read_sentence_and_split_into_array()
print(sentence_array)
print(len(sentence_array))

Enter a sentence: When in the course of human events
['When', 'in', 'the', 'course', 'of', 'human', 'events']
7


In [ ]:
print(sentence_array)
print('number of words in sentence', len(sentence_array))

['When', 'in', 'the', 'course', 'of', 'human', 'events']
number of words in sentence 7


In [ ]:
avg_word2vec = average_word2vec(sentence_array)
print (avg_word2vec)

[0.00048712676, -0.004368693, 0.00017166903, 0.0017402467, 0.0005325078, 0.00057192304, 0.0060858387]


**Cosine similarity** is a measure of similarity between two non-zero vectors of an inner product space based on the cosine of the angle between them, resulting in a value between -1 and 1. A cosine similarity of 1 indicates that the two vectors are pointing in the same direction, while a cosine similarity of -1 indicates that they are pointing in opposite directions. A cosine similarity of 0 indicates that the two vectors are orthogonal.

The cosine similarity function is often used in natural language processing to measure the similarity between two documents, or between a document and a query. It can also be used to measure the similarity between two images.

In [ ]:
### nltk - Natural Language ToolKit
!pip install nltk

In [ ]:
pip install -U sentence-transformers

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 1.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 40.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 23.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 49.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 41.3 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125925 sha256=c90efd8a34e4ac78cb441023d719e77d17312e2035541c584d99a71ec434c517
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence-transformers


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
nltk.download('punkt')
from sentence_transformers import SentenceTransformer, util

# Tokenize the text input.
text = sentence
tokens = nltk.word_tokenize(text)

# Create a vector representation for each token.
vectorizer = TfidfVectorizer()
vectors = vectorizer.fit_transform(tokens)

# Calculate the cosine similarity between the text input and each phrase in the poem dataset.
phrases = X_train
similarities = []

model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')

for phrase in phrases:
    sentence_1= model.encode(sentence, convert_to_tensor=True)
    sentence_2 = model.encode(phrase, convert_to_tensor=True)
    similarity = util.pytorch_cos_sim(sentence_1, sentence_2)
    similarities.append(similarity)

max(similarities)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


tensor([[0.5086]])

In [ ]:
print ('Max cosine Similarity ', max(similarities))
print ('Closest phrase:  ', X_train[similarities.index(max(similarities))])

Max cosine Similarity  tensor([[0.5086]])
Closest phrase:   make a fragrance of her fame.


In [ ]:
print (X_train[similarities.index(max(similarities))])

make a fragrance of her fame.
